### PR TITLE
[GSoC] Added TF and PyTorch detection model conversion cases

### PR DESCRIPTION
**Pull request by LupusSanctus**
GSoC project: "Write a tutorial about training a network in TF or PyTorch and then running it in OpenCV"

**Part 1.**[Classification case](https://github.com/opencv/opencv/pull/17604).
**Part 2.** [Segmentation case](https://github.com/opencv/opencv/pull/17801).
**Part 3.** Detection case.

Provided dnn_model_runner module part for the object detection model retrieval and processing for further reading and inference with OpenCV  API.
Provided tutorial series.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by alalek** [comment]
Please contact with your mentor and check licensing terms of added images. Add `readme.md` files with this information.

File: doc/tutorials/dnn/dnn_pytorch_tf_classification/pytorch_cls_model_conversion_c_tutorial.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+// Check std values.
+if (std.val[0] != 0.0 && std.val[1] != 0.0 && std.val[2] != 0.0)
+{
+    // Divide blob by std.
+    divide(blob, std, blob);
+}
+```
+
+In this step we use [``cv::dnn::blobFromImage(...)``](https://docs.opencv.org/4.4.0/d6/d0f/group__dnn.html#ga29f34df9376379a603acd8df581ac8d7) function to prepare model input.
```
Comment (asmorkalov) [inline doc/tutorials/dnn/dnn_pytorch_tf_classification/pytorch_cls_model_conversion_c_tutorial.md:163]: Please replace the link with link to documentation for master branch, but not particular version.

File: doc/tutorials/dnn/dnn_pytorch_tf_classification/pytorch_cls_model_conversion_tutorial.md [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+## Goals
+In this tutorial you will learn how to:
+* convert PyTorch classification models into ONNX format
+* run converted PyTorch model with OpenCV Python API
+* obtain an evaluation of the PyTorch and OpenCV DNN models
+
+We will explore the above-listed points by the example of the ResNet-50 architecture.
+
+## Introduction
+Let's briefly view the key concepts involved in the pipeline of PyTorch models transition with OpenCV API. The initial step in conversion of PyTorch models into [cv.dnn_Net](https://docs.opencv.org/4.3.0/db/d30/classcv_1_1dnn_1_1Net.html#a82eb4d60b3c396cb85c79d267516cf15)
```
Comment (asmorkalov) [inline doc/tutorials/dnn/dnn_pytorch_tf_classification/pytorch_cls_model_conversion_tutorial.md:12]: The same about link.

File: doc/tutorials/dnn/dnn_pytorch_tf_classification/pytorch_cls_model_conversion_tutorial.md [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+is model transferring into [ONNX](https://onnx.ai/about.html) format. ONNX aims at the interchangeability of the neural networks between various frameworks. There is a built-in function in PyTorch for ONNX conversion: [``torch.onnx.export``](https://pytorch.org/docs/stable/onnx.html#torch.onnx.export).
+Further the obtained ``.onnx`` model is passed into [``cv.dnn.readNetFromONNX``](https://docs.opencv.org/4.3.0/d6/d0f/group__dnn.html#ga7faea56041d10c71dbbd6746ca854197).
```
Comment (asmorkalov) [inline doc/tutorials/dnn/dnn_pytorch_tf_classification/pytorch_cls_model_conversion_tutorial.md:14]: Too.

File: doc/tutorials/dnn/dnn_pytorch_tf_classification/pytorch_cls_model_conversion_tutorial.md [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+
+```python
+# read the image
+input_img = cv2.imread(img_path, cv2.IMREAD_COLOR)
+input_img = input_img.astype(np.float32)
+
+input_img = cv2.resize(input_img, (256, 256))
+
+# define preprocess parameters
+mean = np.array([0.485, 0.456, 0.406]) * 255.0
```
Comment (asmorkalov) [inline doc/tutorials/dnn/dnn_pytorch_tf_classification/pytorch_cls_model_conversion_tutorial.md:125]: It'll be great, if you point the origin of this values and where should users get it for other models.

File: doc/tutorials/dnn/dnn_pytorch_tf_classification/pytorch_cls_model_conversion_tutorial.md [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+
+The below line represents running of the module in the evaluation mode:
+
+```
+python -m dnn_model_runner.dnn_conversion.pytorch.classification.py_to_py_cls --model_name <pytorch_cls_model_name>
+```
+
+Chosen from the list classification model will be read into OpenCV ``cv.dnn_Net`` object. Evaluation results of PyTorch and OpenCV models (accuracy, inference time, L1) will be written into the log file. Inference time values will be also depicted in a chart to generalize the obtained model information.
+
+Necessary evaluation configurations are defined in the [``test_config.py``](https://):
```
Comment (asmorkalov) [inline doc/tutorials/dnn/dnn_pytorch_tf_classification/pytorch_cls_model_conversion_tutorial.md:245]: Link is missing.

Comment (asmorkalov) [reply doc/tutorials/dnn/dnn_pytorch_tf_classification/pytorch_cls_model_conversion_tutorial.md:245]: I propose to add full path to the config file, it's placed in non-obvious location.

File: doc/tutorials/dnn/dnn_pytorch_tf_classification/tf_cls_model_conversion_tutorial.md [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+## Goals
+In this tutorial you will learn how to:
+* obtain frozen graphs of TensorFlow (TF) classification models
+* run converted TensorFlow model with OpenCV Python API
+* obtain an evaluation of the TensorFlow and OpenCV DNN models
+
+We will explore the above-listed points by the example of MobileNet architecture.
+
+## Introduction
+Let's briefly view the key concepts involved in the pipeline of TensorFlow models transition with OpenCV API. The initial step in conversion of TensorFlow models into [cv.dnn_Net](https://docs.opencv.org/4.3.0/db/d30/classcv_1_1dnn_1_1Net.html#a82eb4d60b3c396cb85c79d267516cf15)
```
Comment (asmorkalov) [inline doc/tutorials/dnn/dnn_pytorch_tf_classification/tf_cls_model_conversion_tutorial.md:12]: should be link to master documentation.

File: doc/tutorials/dnn/dnn_pytorch_tf_classification/tf_cls_model_conversion_tutorial.md [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+1. create a TF classification model conversion pipeline and provide the inference
+2. evaluate and test TF classification models
+
+If you'd like merely to run evaluation or test model pipelines, the "Model Conversion Pipeline" tutorial part can be skipped.
+
+### Model Conversion Pipeline
+The code in this subchapter is located in the ``dnn_model_runner`` module and can be executed with the line:
+
+```
+python -m dnn_model_runner.dnn_conversion.tf.classification.py_to_py_mobilenet
```
Comment (asmorkalov) [inline doc/tutorials/dnn/dnn_pytorch_tf_classification/tf_cls_model_conversion_tutorial.md:27]: The procedure requires classes file. I propose to add it to PR or add link to download it.

File: doc/tutorials/dnn/dnn_pytorch_tf_classification/tf_cls_model_conversion_tutorial.md [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+
+The below line represents running of the module in the test mode, namely it provides the steps for the model inference:
+
+```
+python -m dnn_model_runner.dnn_conversion.tf.classification.py_to_py_cls --model_name <tf_cls_model_name> --test True --default_img_preprocess <True/False> --evaluate False
+```
+
+Here ``default_img_preprocess`` key defines whether you'd like to parametrize the model test process with some particular values or use the default values, for example, ``scale``, ``mean`` or ``std``.
+
+Test configuration is represented in [``test_config.py``](https://) ``TestClsModuleConfig`` class:
```
Comment (asmorkalov) [inline doc/tutorials/dnn/dnn_pytorch_tf_classification/tf_cls_model_conversion_tutorial.md:282]: Missing link.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

**Comment by ichisadashioko** [comment]
Is this feature abandoned at the moment?

**Comment by alalek** [comment]
> wants to merge 8 commits

Please squash commits into one, so we can see real size of this patch without history of intermediate changes.

File: doc/tutorials/dnn/dnn_pytorch_tf_classification/pytorch_cls_model_conversion_c_tutorial.md [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+Point classIdPoint;
+double confidence;
+minMaxLoc(prob.reshape(1, 1), 0, &confidence, 0, &classIdPoint);
+int classId = classIdPoint.x;
+```
+
+Here we choose the most likely object class. The ``classId`` result for our case is 335 - fox squirrel, eastern fox squirrel, Sciurus niger.
+In ImageNet ``val.txt`` the correct answer for ``ILSVRC2012_val_00000502.JPEG`` image is 335. Hence, the prediction is correct:
+
+![ResNet50 OpenCV Net C++](images/opencv_resnet50_test_res_c.png)
```
Comment (alalek) [inline doc/tutorials/dnn/dnn_pytorch_tf_classification/pytorch_cls_model_conversion_c_tutorial.md:194]: No need to put lossless images into documentation. They have enormous file size.
It is enough to use JPEG format for that (with quality ~80). File size for VGA image should be less than 150Kb.

**Comment by alalek** [comment]
Please squash commits into one (to trim git history) and rebase on the latest master?

File: doc/tutorials/dnn/dnn_pytorch_tf_classification/pytorch_cls_model_conversion_c_tutorial.md (hunk 1 above)
Comment (alalek) [reply doc/tutorials/dnn/dnn_pytorch_tf_classification/pytorch_cls_model_conversion_c_tutorial.md:163]: > for master branch

Documentation must not refer on docs.opencv.org directly (on any version or master branch).
Use doxygen internal links instead.

**Review by alalek — APPROVED** [review]
